### PR TITLE
Implement basic dynamic calculated field

### PR DESCRIPTION
**Pull request by brandboat**

### CODE REVIEW

File: accio-sqlrewrite/src/main/java/io/accio/sqlrewrite/AccioSqlRewrite.java [hunk 1]
```diff
@@ -54,20 +57,55 @@ public class AccioSqlRewrite
...
-                .addAll(viewDescriptors)
-                .addAll(cumulativeMetricDescriptors)
-                .build();
-        return apply(root, sessionContext, analysis, accioMDL, allDescriptors);
+        Set<QueryDescriptor> allDescriptors;
+        // TODO: Currently DynamicCalculatedField is a experimental feature, and buggy. After all issues are solved,
+        //  we should always enable this setting.
+        if (sessionContext.isEnableDynamicCalculatedField()) {
+            // TODO: make AccioDataLineage static instead of create a new one everytime as it will change only when accio mdl changed.
+            AccioDataLineage dataLineage = AccioDataLineage.analyze(accioMDL);
```
Comment (goldmedal) [inline accio-sqlrewrite/src/main/java/io/accio/sqlrewrite/AccioSqlRewrite.java:80]: How about analyzing data lineage when construct `AccioMDL`? Make `AccioDataLineage` be a member of `AccioMDL`. We don't need to analyze this everytime.

Comment (brandboat) [reply accio-sqlrewrite/src/main/java/io/accio/sqlrewrite/AccioSqlRewrite.java:80]: as TODO mentioned, I'll to this in other PRs.

> Make AccioDataLineage be a member of AccioMDL. We don't need to analyze this everytime.

Currently AccioDataLineage depends on jgrapht and some code in accio-sqlrewrite. If we are going to do this. We will would require some works to do. So as TODO mentioned, I'll open another PR to solve this.

File: accio-sqlrewrite/src/main/java/io/accio/sqlrewrite/AccioDataLineage.java [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    private AccioDataLineage(AccioMDL mdl)
+    {
+        this.mdl = requireNonNull(mdl);
+        this.sourceColumnsMap = collectSourceColumns();
+        this.requiredFieldsByColumn = collectRequiredFieldsByColumn();
+    }
+
+    public LinkedHashMap<String, Set<String>> getRequiredFields(QualifiedName columnNames)
```
[the remark this answers is not in the document]
Comment (brandboat) [reply accio-sqlrewrite/src/main/java/io/accio/sqlrewrite/AccioDataLineage.java]: addressed comments in the latest commit, thanks !

**Review by goldmedal — APPROVED** [review]
LGTM.

**Pull request merged**